Threads.cpp

```diff
@@ -0,0 +1,82 @@
+#include <iostream>
+#include <thread>
+#include <mutex>
+#include <chrono>
+#include <cstdlib>
+#include <vector>
+
+const int CORES = std::thread::hardware_concurrency();
+std::mutex globalLock;
+
+void quickSort(std::vector<int>* arr, int startIdx, int endIdx, int maxthreads = 1) {
+    static int threadsNumber = 1;
+    //std::lock_guard<std::mutex> lock(globalLock);
+
+    if (startIdx != endIdx) {
+        int pointerIdx = startIdx - 1;
+        for (int i = startIdx; i <= endIdx; i++) {
+            if ((*arr)[i] <= (*arr)[endIdx]) {
+                pointerIdx += 1;
+
+                int temp = (*arr)[i];
+                (*arr)[i] = (*arr)[pointerIdx];
+                (*arr)[pointerIdx] = temp;
+            }
+        }
+
+        std::thread left;
+        std::thread right;
+
+        if ((pointerIdx - 1) >= startIdx) {
+            if (threadsNumber < maxthreads) {
+                left = std::thread(quickSort, arr, startIdx, pointerIdx - 1, maxthreads);
```

Comment:
too much threads will be created.
If max threads == 4 -> there will be 16 threads at the end